### PR TITLE
Support toolchain requests with platform-tag style Python implementations and version

### PR DESCRIPTION
**Pull request by zanieb**
Closes https://github.com/astral-sh/uv/issues/4399

### CODE REVIEW

**Comment by zanieb** [comment]
cc @fynnsu

File: crates/uv-toolchain/src/discovery.rs [hunk 1]
```diff
@@ -1268,8 +1275,33 @@ impl FromStr for VersionRequest {
...
+                        major.to_digit(10).map(u8::try_from),
+                        str::parse::<u8>(&minor.iter().collect::<String>()),
+                    ) {
+                        Some(VersionRequest::MajorMinor(major, minor))
+                    } else {
+                        None
+                    }
+                }
+                _ => None,
+            }
```
Comment (zanieb) [inline crates/uv-toolchain/src/discovery.rs:1299]: @BurntSushi I'm embarrassed :)

File: crates/uv-toolchain/src/discovery.rs [hunk 2]
```diff
@@ -1268,8 +1275,33 @@ impl FromStr for VersionRequest {
     type Err = Error;
 
     fn from_str(s: &str) -> Result<Self, Self::Err> {
+        // e.g. `3`, `38`, `312`
+        if let Some(request) = {
```
Comment (BurntSushi) [inline crates/uv-toolchain/src/discovery.rs:1279]: It took me a while to parse this. (With a huge expression on the right hand side of an `if let`.) I'd probably move it up and give it a name?

File: crates/uv-toolchain/src/discovery.rs [hunk 3]
```diff
@@ -1268,8 +1275,33 @@ impl FromStr for VersionRequest {
...
+            match s.chars().collect::<Vec<_>>().as_slice() {
```
Comment (BurntSushi) [inline crates/uv-toolchain/src/discovery.rs:1280]: Hmmm this looks odd to me. I suppose it makes the use of pattern matching nicer below, but I'd probably just work in ASCII land here:

```rust
let bytes = s.as_bytes();
if bytes.len() == 1 {
    if (b'0'..=b'9').contains(&bytes[0]) {
        Some(VersionRequest::Major(bytes[0] - b'0'))
    } else {
        None
    }
}
```

And so on for the `MajorMinor`. If you split this into a helper function, you could probably make it even simpler via early returns.

**Review by BurntSushi — APPROVED** [review]
Nice! I think I would probably write some of this code differently, but I defer to you. :)

**Pull request merged**